crates/matrix-sdk-ui/src/room_list_service/mod.rs

```diff
@@ -323,6 +338,59 @@ impl RoomListService {
         }
     }
 
+    /// Get a [`Stream`] of [`SyncIndicator`].
+    ///
+    /// Read the documentation of [`SyncIndicator`] to learn more about it.
+    pub fn sync_indicator(&self) -> impl Stream<Item = SyncIndicator> {
+        let mut state = self.state();
+
+        stream! {
+            // Ensure the `SyncIndicator` is always hidden to start with.
+            yield SyncIndicator::Hide;
+
+            // Let's not wait for an update to happen. The `SyncIndicator` must be
+            // computed as fast as possible.
+            let mut current_state = state.get();
+
+            loop {
+                let (sync_indicator, yield_delay) = match current_state {
+                    State::Init | State::Recovering | State::Error { .. } => {
+                        (SyncIndicator::Show, SYNC_INDICATOR_DELAY_BEFORE_SHOWING)
+                    }
+
+                    State::SettingUp | State::Running | State::Terminated { .. } => {
+                        (SyncIndicator::Hide, SYNC_INDICATOR_DELAY_BEFORE_HIDING)
+                    }
+                };
+
+                // `state.next().await` has a maximum of `yield_delay` time to execute…
+                let next_state = match timeout(yield_delay, state.next()).await {
+                    // A new state has been received before `yield_delay` time. The new
+                    // `sync_indicator` value won't be yielded.
+                    Ok(next_state) => next_state,
+
+                    // No new state has been received before `yield_delay` time. The
+                    // `sync_indicator` value can be yielded.
+                    Err(_) => {
+                        yield sync_indicator;
+
+                        // Now that `sync_indicator` has been yielded, let's wait on
+                        // the next state again.
+                        state.next().await
+                    }
+                };
+
+                if let Some(next_state) = next_state {
+                    // Update the `current_state`.
+                    current_state = next_state;
+                } else {
+                    // Something is broken with `self.state`. Let's stop this stream too.
+                    break;
+                }
```

Comment:
I think it would be a bit easier to read if this was folded into the `match` above (i.e. one `Ok(Some(next_state))` arm, one `Ok(None)` arm.

Reply:
`next_state` is set in 2 locations. Either from `Ok(next_state) => next_state`, or either in `Err(_) => { …; state.next().await }`. It's actually simpler to manage it here :-p.

Reply:
Oh, I see.